docs/design/paginators.md

```diff
@@ -0,0 +1,297 @@
+
+Smithy services and operations can be marked with the [paginated trait](https://awslabs.github.io/smithy/1.0/spec/core/behavior-traits.html#paginated-trait) which 
+
+> indicates that an operation intentionally limits the number of results returned in a single response an that multiple invocations might be necessary to retrieve all results
+
+
+Pagination works off of a cursor convention where one of the operational input fields is an optional cursor the service 
+uses to produce the next set of results not yet seen. The operation output has a field that marks the cursor position 
+to be used to get the next set of results in subsequent calls.
+
+# Design
+
+Below is an abbreviated example of the trait in use:
+
+```
+service Lambda {
+    operations: [ListFunctions]
+}
+
+@paginated(
+    inputToken: "Marker",
+    outputToken: "NextMarker",
+    pageSize: "MaxItems",
+    items: "Functions"
+)
+@http(method: "PUT", uri: "/2015-03-31/functions", code: 200)
+operation ListFunctions {
+    input: ListFunctionsRequest,
+    output: ListFunctionsResponse
+}
+
+structure ListFunctionsRequest {
+    FunctionVersion: String,
+    Marker: String,
+    MasterRegion: String,
+    MaxItems: Int
+}
+
+structure ListFunctionsResponse {
+    Functions: FunctionConfigurationList,
+    NextMarker: String
+}
+
+list FunctionConfigurationList {
+    member: FunctionConfiguration
+}
+
+structure FunctionConfiguration { ... }
+```
+
+**NOTE**: The trait may be applied at the service level to set default pagination. See the trait definition and examples for the ways it can be used in a model.
+
+
+## Considerations
+
+1. **Discoverability** - pagination is mostly a convenience for common code customers are able to implement themselves. We want them to be able to discover and use the already implemented pagination logic easily so that they don’t end up writing it themselves unnecessarily. 
+2. **Forwards compatibility** - The use of `items` in the `paginated` trait is not required. ~~If a service team later updates the trait to target a specific member using `items` then previous pagination code should continue to work (which means you can’t really generate a paginator for only the member of `items` as adding `items` later would change the generated pagination code in a way that isn’t backwards compatible).~~
+    1. Update 2/12/2021: The Smithy team has indicated this would not be a valid model evolution. The recommendation was still to provide a way to paginate over the raw responses as well as providing a flatMap like operation for the `items` if specified.
+
+
+## API
+
+The SDK runtime will provide a `SdkAsyncIterable` interface that represents the public API for a paginated operation.
+
+
+```kotlin
+/**
+ * An asynchronous data stream. This type is an asynchronous version of [Iterable].
+ */
+interface SdkAsyncIterable<out T> {
+   operator fun iterator(): SdkAsyncIterator<T>
+}
+
+/**
+ * An asynchronous [Iterator].
```

Comment:
suggestion

`An [Iterator] which asynchronously returns the next value` since it doesn't look like `hasNext()` suspends.